extensions/sql-database-projects/src/dialogs/deployDatabaseDialog.ts

```diff
@@ -0,0 +1,291 @@
+		};
+	}
+
+	private createDataSourcesDropdown(view: azdata.ModelView): azdata.FormComponent {
+		let dataSourcesValues: DataSourceDropdownValue[] = [];
+
+		this.project.dataSources.forEach(dataSource => {
+			const dbName: string = (dataSource as SqlConnectionDataSource).getSetting('Initial Catalog');
+			const displayName: string = `${dataSource.name}`;
+			dataSourcesValues.push({
+				displayName: displayName,
+				name: dataSource.name,
+				dataSource: dataSource,
+				database: dbName
+			});
+		});
+
+		this.dataSourcesDropDown = view.modelBuilder.dropDown().withProperties({
+			values: dataSourcesValues,
+		}).component();
+
+
+		this.dataSourcesDropDown.onValueChanged(() => {
+			this.setDatabaseToSelectedDataSourceDatabase();
+			this.tryEnableGenerateScriptAndOkButtons();
+		});
+
+		return {
+			title: constants.dataSourceDropdownTitle,
+			component: this.dataSourcesDropDown
+		};
+	}
+
+	private setDatabaseToSelectedDataSourceDatabase(): void {
+		if ((<DataSourceDropdownValue>this.dataSourcesDropDown!.value).database) {
+			this.targetDatabaseTextBox!.value = (<DataSourceDropdownValue>this.dataSourcesDropDown!.value).database;
+		}
+	}
+
+	private createEditConnectionButton(view: azdata.ModelView): azdata.Component {
+		let editConnectionButton: azdata.ButtonComponent = view.modelBuilder.button().withProperties({
+			label: constants.editConnectionButtonText,
+			title: constants.editConnectionButtonText,
+			ariaLabel: constants.editConnectionButtonText
+		}).component();
+
+		editConnectionButton.onDidClick(async () => {
+			this.connection = <azdata.connection.ConnectionProfile><any>await azdata.connection.openConnectionDialog();
+
+			// show connection name if there is one, otherwise show connection string
+			if (this.connection.options['connectionName']) {
+				this.targetConnectionTextBox!.value = this.connection.options['connectionName'];
+			} else {
+				this.targetConnectionTextBox!.value = await azdata.connection.getConnectionString(this.connection.connectionId, false);
+			}
+
+			// change the database inputbox value to the connection's database if there is one
+			if (this.connection.options.database) {
+				this.targetDatabaseTextBox!.value = this.connection.options.database;
+			}
+		});
+
+		return editConnectionButton;
+	}
+
+	private createClearButton(view: azdata.ModelView): azdata.Component {
+		let clearButton: azdata.ButtonComponent = view.modelBuilder.button().withProperties({
+			label: constants.clearButtonText,
+			title: constants.clearButtonText,
+			ariaLabel: constants.clearButtonText
+		}).component();
+
+		clearButton.onDidClick(() => {
+			this.targetConnectionTextBox!.value = '';
+		});
+
+		return clearButton;
+	}
+
+	// only enable Generate Script and Ok buttons if all fields are filled
+	private tryEnableGenerateScriptAndOkButtons(): void {
+		if (this.targetConnectionTextBox!.value && this.targetDatabaseTextBox!.value && this.deployScriptNameTextBox!.value
```

Comment:
Why the non-null assertion here?

Reply:
It's to get rid of the "Object is possibly undefined" error because the types of these guys is also undefined because they don't get assigned in the constructor

Reply:
That's fair, as long as you're absolutely certain they won't be undefined by the time you're in this method :)